peps/pep-0734.rst

```diff
@@ -0,0 +1,800 @@
+
+An ``interpreters.Interpreter`` object represents the interpreter
+(``PyInterpreterState``) with the corresponding ID.
+
+If the interpreter was created with ``interpreters.create()`` then
+it will be destroyed as soon as all ``Interpreter`` objects have been
+deleted.
+
+Attributes and methods:
+
+* ``id``
+      (read-only) The target interpreter's ``InterpreterID`` object.
+      It is an int-like object with some internal bookkeeping.
+
+* ``__hash__()``
+      Returns the hash of the interpreter's ``id``.  This is the same
+      as the hash of the ID's integer value.
+
+* ``__eq__(other)``
+      Returns ``other is self``.
+
+* ``is_running() -> bool``
+      Returns ``True`` if the interpreter is currently executing code
+      in its ``__main__`` module.  This excludes sub-threads.
+
+      It refers only to if there is a thread
+      running a script (code) in the interpreter's ``__main__`` module.
+      That basically means whether or not ``Interpreter.exec()`` is
+      running in some thread.  Code running in sub-threads is ignored.
+
+* ``prepare___main__(**kwargs)``
+      Bind one or more objects in the interpreter's ``__main__`` module.
+
+      The keyword argument names will be used as the attribute names.
+      The values will be bound as new objects, though exactly equivalent
+      to the original.  Only objects specifically supported for passing
+      between interpreters are allowed.  See `Shareable Objects`_.
+
+      ``prepare___main__()`` is helpful for initializing the
+      globals for an interpreter before running code in it.
+
+* ``exec(code, /)``
+      Run the given source code in the interpreter
+      (in the current thread).
+
+      This is essentially equivalent to switching to this interpreter
+      in the current thread and then calling the builtin ``exec()``
+      using this interpreter's ``__main__`` module's ``__dict__`` as
+      the globals and locals.
+
+      ``exec()`` does not reset the interpreter's state nor
+      the ``__main__`` module, neither before nor after, so each
+      successive call picks up where the last one left off.  This can
+      be useful for running some code to initialize an interpreter
+      (e.g. with imports) before later performing some repeated task.
+
+Comparison with builtins.exec()
+-------------------------------
+
+``Interpreter.exec()`` is essentially the same as the builtin
+``exec()``, except it targets a different interpreter, using that
+interpreter's distinct runtime state.
+
+Here are the relevant characteristics of the builtin ``exec()``,
+for comparison:
+
+* It runs in the current OS thread and pauses whatever was running there,
+  which resumes when ``exec()`` finishes.  No other threads are affected.
+  (To avoid pausing the current thread, run ``exec()``
+  in a ``threading.Thread``.)
+* It executes against a namespace, by default the ``__dict__`` of the
+  current module (i.e. ``globals()``).  It uses that namespace as-is
```

Comment:
The default for builtins.exec() is not always globals() -- when run inside a function it is locals(). I think Interpreter.exec() always runs against the namespace of *that interpreter's* `__main__` module, right?

Reply:
Good catch!